packages/fhir-sanitize-param/index.js

```diff
@@ -72,142 +72,230 @@ function findMatchForConfig(name, params) {
+ * @description Sanitize token values
+ * @param name
+ * @param inputValue
+ * @param type
+ * @returns {*}
+ */
+function sanitizeToken(name, inputValue, type = 'token') {
+	// Throw if the value is not a string, because we will not be able to make a token out of it
+	invariant(typeof inputValue === 'string', mismatchError(type, name));
+	let chunks = inputValue.split('|');
+
+	// Tokens have 1 or 2 parts containing codes and systems that are separated by pipes.
+	invariant([1, 2].includes(chunks.length), mismatchError(type, name));
+	let system;
+	let code;
+	switch (chunks.length) {
+		case 1:
+			system = '';
+			code = chunks[0];
 			break;
-		case 'json_string':
-			// This is a custom field for writing values, we need to accept
-			// strings of JSON and attempt to parse them. There will also
-			// need to be validation done at the write level to make sure the
-			// resource is itself valid
-			// This will throw if value is not correct, no need to invariant
-			result = JSON.parse(value);
+		case 2:
+			system = chunks[0];
+			code = chunks[1];
 			break;
-		default:
-			// Force an invariant here
-			invariant(false, unsupportedError(type, name));
 	}
+	system = validator.stripLow(xss(sanitize(system)));
+	code = validator.stripLow(xss(sanitize(code)));
 
-	return result;
+	return {
+		code: code,
+		system: system
+	};
 }
 
 /**
- * @function parseArguments
- * @description Parse only arguments needed for this type of request
- * @param {Express.req} req - Request from an express server
- * @return {Object} - Arguments object
+ * @function sanitizeQuantity
+ * @description Sanitize quantity values
+ * @param name
+ * @param inputValue
+ * @param type
+ * @returns {*}
  */
-function parseArguments(req) {
-	let args = {};
-	// Merge query from get only
-	if (req.method === 'GET') {
-		args = Object.assign(args, req.query);
-	}
-	// Merge body from put and post only
-	if (/PUT|POST/.test(req.method)) {
-		args = Object.assign(args, req.body);
+function sanitizeQuantity(name, inputValue, type = 'quantity') {
+	let [number, token] = inputValue.split(/\|(.+)/);
+	number = sanitizeNumber(name, number, 'quantity.number');
+	let system = '';
+	let code = '';
+	if (token) {
+		({ system, code } = sanitizeToken(name, token, 'quantity.token'));
 	}
-	// Merge params for all requests
-	return Object.assign(args, req.params);
+	invariant(number, mismatchError(type, name));
+
+	return {
+		number: number,
+		system: system,
+		code: code,
+	};
+}
+
+/**
+ * @function coerceValue
+ * @description Given a value and a type, attempt to coerce the value to the correct type
+ * @param config
+ * @param requestValue
+ * @param isMissingQuery
+ * @returns {*}
+ */
+function coerceValue(config, requestValue, isMissingQuery) {
+	let { name, type } = config;
+	requestValue = Array.isArray(requestValue) ? requestValue : [requestValue];
+	const sanitizedResult = requestValue.map(values => {
+		invariant(typeof values === 'string', mismatchError(type, name));
```

Comment:
aren't values always strings?

Reply:
I think they do actually always come in as strings from express, this is just being defensive.